### PR TITLE
Rule S1147: Exit methods should not be called

### LINKED ISSUE #1071: Rule S1147: Exit methods should not be called
[RSPEC-1147](https://jira.sonarsource.com/browse/RSPEC-1147)

Calling Environment.Exit(exitCode) or Application.Exit() terminates the process and returns an exit code to the operating system..

### PR DESCRIPTION
**Pull request by michalb-sonar**
Fix #1071

### CODE REVIEW

**Comment by michalb-sonar** [comment]
No idea what's the thing with new lines again. I have autocrlf set to `false`...
For more readable diff, append `?w=1` to the url.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotCallExitMethods.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Windows.Forms;
+
+namespace Tests.Diagnostics
+{
+    class Program
+    {
+        public void Foo()
+        {
+            Environment.Exit(0); // Noncompliant
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotCallExitMethods.cs]: You forgot the cases specified in the exclude (i.e. the rule should not report when called in the main method).

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotCallExitMethods.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+using System.Windows.Forms;
+
+namespace Tests.Diagnostics
+{
+    class Program
+    {
+        public void Foo()
+        {
+            Environment.Exit(0); // Noncompliant
+            System.Windows.Forms.Application.Exit(); // Noncompliant
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotCallExitMethods.cs]: Same here.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallExitMethods.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        protected override bool ShouldReportPreCheck(InvocationExpressionSyntax invocationSyntax,
+            SemanticModel semanticModel)
+        {
+            var parent = invocationSyntax.Parent;
+            while (parent != null)
+            {
+                if (parent is BaseMethodDeclarationSyntax)
+                {
+                    var parentMethodSymbol = semanticModel.GetDeclaredSymbol(parent) as IMethodSymbol;
+                    return !parentMethodSymbol.IsMainMethod();
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallExitMethods.cs]: I think this doesn't work with inner methods (C# 7).

Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallExitMethods.cs]: Good catch. It works, but I had to check. Added unit tests.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallMethodsBase.cs [hunk 4]
```diff
@@ -48,32 +48,39 @@ private void AnalyzeInvocation(SyntaxNodeAnalysisContext analysisContext)
...
 
             if (ShouldReportOnMethodCall(invocation, analysisContext.SemanticModel))
             {
                 analysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, identifier.Value.GetLocation(),
-                    methodSignature.ToShortName()));
+                    disallowedMethodSignature.ToShortName()));
             }
         }
 
+        protected virtual bool ShouldReportPreCheck(InvocationExpressionSyntax invocationSyntax,
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallMethodsBase.cs]: I would rename this method to something that express better the fact this is done before checking the method call (e.g. `IsInValidContext()`).

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallExitMethods.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        {
+            new MethodSignature(KnownType.System_Environment, "Exit"),
+            new MethodSignature(KnownType.System_Windows_Forms_Application, "Exit")
+        };
+        internal override IEnumerable<MethodSignature> CheckedMethods => checkedMethods;
+
+        protected override bool ShouldReportPreCheck(InvocationExpressionSyntax invocationSyntax,
+            SemanticModel semanticModel)
+        {
+            var parent = invocationSyntax.Parent;
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallExitMethods.cs]: Why not use the `.Ancestors` property and so LINQ?

**Pull request merged**